### ISSUE TITLE
Injecting JSON-LD data into Dataset Landing page

### ISSUE THREAD

**Issue by rushirajnenuji** [NCEAS/metacatui#482]
As concluded from issue #432  that Google bot can parse structured `JSON-LD` data from a dynamically generated webpage, next steps would be to use `SOLR` data and inject it into dataset landing page.

**Comment by rushirajnenuji** [comment]
While injecting the `JSON-LD` object into the DOM,
```
<script type="application/jsonld">
.
.
</script>
```
- The JSON-LD data gets injected correctly while rendering the first dataset.

- Exceptions arise when the application has to render another dataset view after the first one.
The issue is that the JSON-LD data gets appended as a new script with `type=" application/jsonld"` to the DOM.

- Assigning `id` to the script type and checking for already existing scripts with type `application/jsonld` solves the problem and correctly replaces the LinkedData for the newly rendered page.

- My question is: Will the browser be aware of this change and parse the linked data again?!
- If not, what would be the correct way to handle this issue?!

**Comment by rushirajnenuji** [comment]
Hey @NCEAS/dev,
@amoeba and I, have been working on trying to inject JSON linked data into our Catalog and Dataset landing pages.
- Our approach: use the `solr` information available and populate that information dynamically on the front end.
- So far, we've been successful in doing so for the majority of the attributes for the schema found at [Schema.org/Dataset](http://schema.org/Dataset) and [Project418/Vocab](https://github.com/earthcubearchitecture-project418/p418Vocabulary/tree/master/html/voc/static/schema).
- However, some of the information isn't directly available from `solr` and we might get by parsing the `eml` which could result in possible delays while rendering information on the front end.

Thoughts / Suggestions?!

**Comment by amoeba** [comment]
We talked on our weekly geolink call today about nailing down all of these properties and decided it would be great to make a spreadsheet for all the properties of the  Repository, DataCatalog, and Dataset classes that'll help us decide and plan for publishing JSON-LD for them. Make a Google Sheet with columns like property, whether to publish the property or not, whether the information is in the Solr index currently, could be, or has to be extracted from the metadata. Probably some more fields I'm not thinking of.

**Comment by amoeba** [comment]
On our call last week, it was suggested to just put together a spreadsheet to help discuss and decide on which properties to want to publish. I made a first attempt at it here https://docs.google.com/spreadsheets/d/1jQpzOruUEdP0rcrwkyf7XcQVVLVbx6ofU0porXQFp0k/edit?usp=sharing. I'm going to go over it in the morning because I know a missed a thing or two but we should talk about it soon.

**Comment by amoeba** [comment]
This is looking really good: https://handy-owl.nceas.ucsb.edu/view/test.10.1

I'd still love to see the handy owl metacatui pointed to dataone.org so we can run through a wide variety of content to see if everything works (no JS runtime errors) and is producing the JSON-LD we want.

**Comment by amoeba** [comment]
Talked on dev call: Decided to schedule this for the next release. Blockers for deploying this are:

- Finalizing Dataset and DataCatalog publishing pattern
- Testing the code itself by pointing a test instance at dataone.org and testing a large variety of content
- Deciding on a PURI-space for Datasets and DataCatalog so our `@id`s are set in stone

**Comment by amoeba** [comment]
I rounded out the existing code with some extra features in 310178bccd4669ac64f4f461d1eb0d8f934544e8. Did some testing w/ dev.nceas today and it looks good. See the commit for what was added.

Merging this is still blocked by:

- Deciding on a URI space for the `@id` (URI) of the dataset resource: https://redmine.dataone.org/issues/8601
- Testing in a CN environment (That's tomorrow's goal)

**Comment by amoeba** [comment]
Still making lots of progress on this by doing some random testing on content from DataONE search and finding a lot of minor bugs. I'll wrap up testing tomorrow.

**Comment by amoeba** [comment]
Tested JSONLD generation by dumping the JSONLD for two datasets from each of KNB, LTER, ARCTIC, NCEI, PANGAEA. I tested both a recent and old doc from each, covering EML211, eml-dataset (module level schema), ISO 19115 (normal, noaa, and pangaea variants). Found a good handful of bugs with my code as well as bugs in DataONE indexing (filed). I also linted the GeoJSON for all my tests and found all the GeoJSON to be valid.

Still blocking:

- [ ] Deciding on a URI space for the @id (URI) of the dataset resource: https://redmine.dataone.org/issues/8601
- [ ] Deploy and test on a test MN and CN environment
